### PR TITLE
feat(orchestrator): init Russula framework

### PR DESCRIPTION
**Pull request by toidiu**
https://github.com/aws/s2n-netbench/issues/21

### Suggestions on how to review this PR
One suggestion to help review this PR would be to read the entire PR description and Readme in the PR to gain an understanding of what Russula is and why we need it.

Following this it could be useful to understand how Russula will be used. Unfortunately, this is harder to decipher since that will be introduced in following PRs. See https://github.com/aws/s2n-netbench/issues/21 for a list of upcoming PRs. Instead one can look here [link](https://github.com/aws/s2n-netbench/blob/ak-fullOrchImport/netbench-orchestrator/src/orchestrator.rs#L180-L203) to see how it will finally be used in the orchestrator.

Reviewing files:
- mod.rs: the entry point which contains the builder and the russula::Endpoint. The endpoint will be entry point for application and exposes 2 public APIs.
- protocol.rs: the underlying mechanism for Russula. Protocol is a trait and impl of Protocol will be introduced in following PRs. The most important fn in Protocol is `poll_state_impl`, which essentially checks if we are the desired state, otherwise `run`s actions associated with the current state.
- state.rs: `trait Protocol` has an associated type Sate which is used to reason about state transitions. This file captures some common functionality associated with State.
- network_utils.rs: has usesful fn to send/recv messages between russula endpoints. The file also includes the struct Msg is composed of len and data (essentially [TLV](https://en.wikipedia.org/wiki/Type%E2%80%93length%E2%80%93value) encoding but minus the tag). Msg is used to read a single message at a time, in a serialized manner, from the socket since messages have side-effects depending on the current state and can also lead to state transitions.
- other files: should be pretty self explanatory

### Description
This PR creates the netbench-orchestrator project and introduces the Russula component.

**An overview of Russula**
Russula is component which will be used by netbench-orchestrator (coming in following PRs) to coordinate netbench servers and clients on remote hosts. Russula is composed of a pair of Coordinator/Worker instances. A single Russula Coordinator is used to coordinate and drive multiple Russula Workers.

**Why do we need to coordinate netbench clients and servers?**

To answer why we need to coordinate clients/server, lets look at what is involved when running netbench manually: 
- generate a netbench scenario file
- launch a netbench server (provide the scenario file)
- **wait** for netbench server to start running 
- launch a netbench client (provide the same scenario file) + (provide a list of server addresses)
- **wait** for the netbench client to finish (recv/send all data which can vary based on environment)
- manually stop the netbench server
- generate a netbench report

The difficulty when trying to automate the above steps is to account for the **wait**s. Wait too little and the netbench run will fail. Run too long (but how long?) and the report will include alot of uninteresting data (potentially making the report difficult to parse). Throw in the variability of network delays/loss, variability of the cloud, running multiple netbench drivers in a single run, and running incase scenarios with multiple servers/clients; its easy to see that naively waiting is not the correct solution. In fact earlier iterations of the project validated that waiting naively is not the correct solution.

**Coordinating** the servers/clients essentially means that some process has direct visibility into how long we should wait and can execute the next step as soon as the previous step has completed.

### Call-outs:
This simply introduces the Russula framework. Following PRs introduce netbench specific Russula protocols, which are used to test the protocol.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: netbench-orchestrator/src/russula/event.rs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+impl EventRecorder {
+    pub fn process(&mut self, event: EventType) {
+        match event {
+            EventType::SendMsg => self.send_msg += 1,
+            EventType::RecvMsg => self.recv_msg += 1,
+        }
+    }
+}
+
+impl Display for EventRecorder {
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/event.rs:31]: Could we just stick with "Debug" for this? It looks like the display implementation is essentially the same thing.

File: netbench-orchestrator/src/russula/mod.rs [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+            if !peer.protocol.[< is_ $state _state>]() {
+                return false;
+            }
+        }
+        true
+    }
+}};
+}
+
+impl<P: Protocol + Send> Russula<P> {
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/mod.rs:89]: I wonder if you could avoid the macros here by using `enums` for the states instead?

E.g. instead of
```
is_ready_state()
is_done_state()
is_woker_running_state()
```

You could just have
```
is_state(ProtocolState::Ready)
```
Or something like that.

File: netbench-orchestrator/src/russula/protocol.rs [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    /// Track events for the current protocol.
+    fn event_recorder(&mut self) -> &mut EventRecorder;
+
+    /// Used for debugging and creating unique log files.
+    fn name(&self) -> String;
+
+    /// Track the peers state. Mainly used for debugging.
+    fn update_peer_state(&mut self, msg: Msg) -> RussulaResult<()>;
+
+    state_api!(ready);
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/protocol.rs:86]: Similar "enum over macro" as above.

File: netbench-orchestrator/src/russula/network_utils.rs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+use tokio::io::AsyncWriteExt;
+use tokio::net::TcpStream;
+use tracing::error;
+
+// Msg.len is represented as a u16
+const LEN_PREFIX_BYTES: usize = 2;
+
+/// Data format for messages exchanged with a peer.
+#[derive(Debug)]
+pub struct Msg {
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/network_utils.rs:16]: If feels like we're reinventing the wheel a bit here. Instead of dealing with the encoding ourselves, could we instead define a semantic message type (which represents the meaning of a message) and then just use serde for serialization/deserialization?

File: netbench-orchestrator/src/russula/states.rs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+}
+
+pub trait StateApi: Debug + Serialize + for<'a> Deserialize<'a> {
+    /// The [TransitionStep] required to move to the next state.
+    fn transition_step(&self) -> TransitionStep;
+
+    /// Returns the next step in the state machine.
+    fn next_state(&self) -> Self;
+
+    fn eq(&self, other: &Self) -> bool {
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/states.rs:39]: Could we instead specify a bound on the `Equal` trait or whatever it's called?

File: netbench-orchestrator/src/russula/mod.rs [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+        Self {
+            protocol_addr_pair_list: peer_list,
+            poll_delay,
+        }
+    }
+
+    pub async fn build(self) -> RussulaResult<Russula<P>> {
+        let mut stream_protocol_list = Vec::new();
+        for (addr, protocol) in self.protocol_addr_pair_list.into_iter() {
+            let mut retry_attempts = 10;
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/mod.rs:136]: Nit: magic number, could we use a constant instead? I think I saw a "3 for package loss" some other place too.

File: netbench-orchestrator/src/russula/mod.rs [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+                            addr,
+                            stream: connect,
+                            protocol,
+                        });
+
+                        break;
+                    }
+                    Err(err) => {
+                        error!(
+                            "Failed to connect.. wait and retry. Try disabling VPN and check your network connectivity.
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/mod.rs:156]: Nit: If it failed 10 times, do we actually expect "wait and retry" to be a valid strategy? 😛

File: netbench-orchestrator/src/russula/protocol.rs [hunk 8]
```diff
@@ -0,0 +1,252 @@
...
+    }
+
+    $(#[$meta])*
+    async fn [<poll_ $state>](&mut self, stream: &mut TcpStream) -> RussulaResult<Poll<()>> {
+        self.poll_state(stream, &self.[<$state _state>]()).await
+    }
+}};
+}
+
+macro_rules! notify_peer {
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/protocol.rs:45]: Could this be a function with generic arguments rather than a macro?

**Review by jmayclin — COMMENTED** [review]
Completed an initial pass. Let me know your thoughts on switching out macros for some enum oriented things. I'm hoping that will make it easier for me to read things, and will do a second pass then.

File: netbench-orchestrator/src/russula/network_utils.rs (hunk 4 above)
Comment (toidiu) [reply netbench-orchestrator/src/russula/network_utils.rs:16]: Not sure I understand the suggestion here. Could you elaborate on what you mean.

File: netbench-orchestrator/src/russula/mod.rs (hunk 7 above)
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs:156]: A Coordinator attempts to connect to multiple workers. If the Coordinator is launched after the Worker, then retry can catch the failure.

Depending on the `poll_delay` the user passes in, the retries will run fast or slow. In practice I dont expect us to get to 10 very often.

File: netbench-orchestrator/src/russula/mod.rs (hunk 2 above)
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs:89]: yea macros can be annoying to read. lemme try and see how using enum looks like

File: netbench-orchestrator/src/russula/states.rs (hunk 5 above)
Comment (toidiu) [reply netbench-orchestrator/src/russula/states.rs:39]: I should note that I specifically do not want to derive Eq/PartialEq but instead compare the `serde_json.to_string` because I want to ignore some fields (next PR) when doing the eq comparison.

For example I dont want to compare the u32 which represents a PID:
`Running(#[serde(skip)] u32)`

File: netbench-orchestrator/src/russula/mod.rs [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+    pub stream: TcpStream,
+    pub protocol: P,
+}
+
+/// A Russula Endpoint.
+///
+/// An Endpoint can be of type Coordinator or Worker. A Coordinator can
+/// be used to synchronize multiple workers across different hosts. A Worker
+/// communicates with a Coordinator to make progress.
+pub struct Endpoint<P: Protocol> {
```
Comment (jmayclin) [inline netbench-orchestrator/src/russula/mod.rs:49]: It looks like this top level Endpoint struct has to be pretty involved with the low level details of polling all of the little downstream things. It there a way that we could push all of the polling details out of this struct?

I'm thinking that one appraoch might be to leverage the inherent "event passing" notification that you get when data is delivered to a socket with tokio. E.g. currently we
1. receive data on a TCP Stream
2. wait until some top level thing manually calls poll to update the state.

Rather than waiting to manually update things, could you just spawn a tokio task for each TCP Stream so that the state is automatically updated whenever you receive data in the socket? My hope is that this would make polling/updating the overall state significantly easier.

File: netbench-orchestrator/src/russula/network_utils.rs (hunk 4 above)
Comment (toidiu) [reply netbench-orchestrator/src/russula/network_utils.rs:16]: I already use [serde_json](https://github.com/aws/s2n-netbench/pull/29/files?diff=unified&w=1#diff-86f35409967dedc6617df9b08c5d65fbc0217b7be58c401245a052c22409f628R44-R49) to generate the payload.

The goal of Msg is to read exactly a Msg amount of bytes from the socket. In order to do that we need to prepend `len` to the payload and we need to only read `len` amount of bytes from the socket.

File: netbench-orchestrator/src/russula/mod.rs (hunk 7 above)
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs:149]: Figuring out that the error was happening due to VPN was not obvious so I dont think it hurts to leave it in there. If the user has VPN disabled then they can explore other solutions. I can clarify that disabling VPN could be one solution.

File: netbench-orchestrator/src/russula/mod.rs [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+impl<P: Protocol + Send> Endpoint<P> {
+    pub async fn run_till(&mut self, state: ProtocolState) -> RussulaResult<()> {
+        while self.poll_state(state).await?.is_pending() {
+            tokio::time::sleep(self.poll_delay).await;
+        }
+
+        Ok(())
+    }
+
+    pub async fn poll_state(&mut self, state: ProtocolState) -> RussulaResult<Poll<()>> {
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs:66]: There are two ways to interact with a Russula endpoint: `poll_state` or `run_till`. Poll is non-blocking while run_till blocks. Would a comment help or were you looking for something else?

File: netbench-orchestrator/src/russula/protocol.rs [hunk 11]
```diff
@@ -0,0 +1,246 @@
...
+    states::{StateApi, TransitionStep},
+    ProtocolState, RussulaResult,
+};
+use crate::russula::event::EventRecorder;
+use core::{task::Poll, time::Duration};
+use std::net::SocketAddr;
+use tokio::net::TcpStream;
+use tracing::{debug, info};
+
+// Send the Done status multiple times to the peer in case there is packet loss.
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/src/russula/protocol.rs:18]: The goal was to not rely on the transport.. because we might replace TCP with something else. The main purpose of this mechanism is to send the terminal state more than once to the peer since we are shutting down our side and wont be interacting with the peer after this. Besides packet_loss there are other reason in a distributed system why this is necessary (delay)

I was glossing over all this detail but can see why `packet_loss` is a confusing term. I'll remove it and simplify the wording.

File: netbench-orchestrator/src/russula/mod.rs [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+            if !peer.protocol.is_state(state) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+type SockProtocol<P> = (SocketAddr, P);
+pub struct RussulaBuilder<P: Protocol> {
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs:101]: Russula is agnostic over netbench. In future PRs we will see that you can implement Netbench specific protocols in Russula so I would like to keep the naming separate. There is a description of what Russula is in the Readme. If some part of that is not clear then I can expand on that.

File: netbench-orchestrator/src/russula/protocol.rs [hunk 13]
```diff
@@ -0,0 +1,246 @@
...
+const NOTIFY_DONE_TIMEOUT: Duration = Duration::from_secs(1);
+// Notify done multiple time in case of packet loss.. this is best effort
+const DONE_SENT_COUNT: usize = 3;
+
+pub trait Protocol: Clone {
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/src/russula/protocol.rs:23]: Protocol has no state and generic over objects that impl the protocol trait. Objects that implement Protocol will have state See the associated type (State) in Protocol.

File: netbench-orchestrator/src/russula/network_utils.rs (hunk 4 above)
Comment (jmayclin) [reply netbench-orchestrator/src/russula/network_utils.rs:16]: > The goal of Msg is to read exactly a Msg amount of bytes from the socket

I agree that the Msg format accomplishes that, but it seems like the purpose of doing that is "we need a way to read json structs from the wire and deserialize them into Rust objects". My hope is that there is a way to do that without needing to write our own length-encoding scheme. My favorite code is no code at all 😄

Comment (WesleyRosenblum) [reply netbench-orchestrator/src/russula/network_utils.rs:16]: I agree with @jmayclin on this. I think you should be able to accomplish your goal of sending and receiving multiple messages through existing methods such as serde. For example, couldn't you just serialize a vec of messages and deserialize them on the other end, and then process them one by one? 

I think its nice you've gone to such lengths to define the message encoding, etc, but I could only really see that necessary in an absolutely performance critical implementations like `s2n-quic`. For Netbench we can tolerate the tiny bit of overhead that may be introduced by existing solutions, so I think its worth trying to pare things down to just the essential pieces that are unique to Netbench.

Comment (toidiu) [reply netbench-orchestrator/src/russula/network_utils.rs:16]: > but I could only really see that necessary in an absolutely performance critical implementations

`Msg` doesn't add performance, it adds the ability to read 1 and only 1 payload at a time from the wire. That single payload can be json,txt,binary encoded. In the currently impl it is json encoded using the serde_json library.

> "we need a way to read json structs from the wire and deserialize them into Rust objects". My hope is that there is a way to do that without needing to write our own length-encoding scheme

I am not sure what a different solution looks like.

> For example, couldn't you just serialize a vec of messages and deserialize them on the other end, and then process them one by one?

What is 'message' here?

File: netbench-orchestrator/README.md [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+## Getting started
+
+**Pre-requisites**
+- Built and include [netbench](https://github.com/aws/s2n-netbench) utilities (`cargo build`)
+  - Include in PATH `export PATH="s2n-netbench/target/release/:$PATH"`. Test with `which s2n-netbench`
+- AWS cli is installed. Test with `which aws`
+- An AWS account with some infrastructure configured. TODO: provide an easy way to do this
+  - Make sure AWS credentials are included in your shell environment
+- The ec2 SSH key name is correctly set in state.rs (make this configurable)
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/README.md:18]: It gets introduced in a later PR, is only necessary for debugging and will prob default to None. I can just remove this line.

https://github.com/aws/s2n-netbench/blob/ak-fullOrchImport/netbench-orchestrator/src/orchestrator/state.rs#L30

File: netbench-orchestrator/README.md [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+support. Logs are written to a file `orch_proj/target/russula.log*` file on the host. The
+`make run_orchestrator` command enables sane log levels via `RUST_LOG=...` but these can be
+changed as desired.
+
+#### Remote
+**SSH access**
+ec2 accepts the name of a ssh-key when creating a new host. This is set to a default value
+under the value `ssh_key_name` in the [state.rs](/src/state.rs) file. By providing this key
+it is possible to ssh onto the remote host locally: `ssh -oStrictHostKeyChecking=no ec2-user@x.x.x.x`.
+Its also possible to ssh onto a host from the ec2 console on AWS.
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/README.md:60]: We will specifically be launching al2023 so ec2-user should be accurate. The orchestrator relies on ssm so it will prob take some effort to run on other platforms. https://github.com/aws/s2n-netbench/blob/ak-fullOrchImport/netbench-orchestrator/src/orchestrator/state.rs#L28

File: netbench-orchestrator/README.md [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+#### Local
+**Orchestrator**
+The Orchestrator is Rust code and ships with [tracing](https://docs.rs/tracing/latest/tracing/)
+support. Logs are written to a file `orch_proj/target/russula.log*` file on the host. The
+`make run_orchestrator` command enables sane log levels via `RUST_LOG=...` but these can be
+changed as desired.
+
+#### Remote
+**SSH access**
+ec2 accepts the name of a ssh-key when creating a new host. This is set to a default value
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/README.md:57]: yep, the latest it is. I can cleanup this comment

File: netbench-orchestrator/src/russula/mod.rs [hunk 17]
```diff
@@ -0,0 +1,180 @@
...
+        // If the peer is already in the desired state then this should be a noop.
+        for peer in self.instance_list.iter_mut() {
+            if let Err(err) = peer
+                .protocol
+                .poll_state(&mut peer.stream, desired_state)
+                .await
+            {
+                if err.is_fatal() {
+                    error!("{} {}", err, peer.addr);
+                    panic!("{} {}", err, peer.addr);
```
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs:84]: Since the error is fatal, there is not much the application can do about it. An application would need to check if the error is fatal and shutdown/panic.

But its prob better to let the application handle it rather than panic.

File: netbench-orchestrator/src/russula/network_utils.rs (hunk 4 above)
Comment (WesleyRosenblum) [reply netbench-orchestrator/src/russula/network_utils.rs:16]: > What is 'message' here?

Something higher level than what you have here, which looks like just some raw bytes. I would expect the message to be the actual instruction you want to send, probably an enum.

File: netbench-orchestrator/src/russula/mod.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (toidiu) [reply netbench-orchestrator/src/russula/mod.rs]: The "establish a usable connection during the build step" is not uncommon and we use it in s2n-quic. A [Client::connect](https://docs.rs/s2n-quic/latest/s2n_quic/client/struct.Client.html#method.connect) returns a [Connection](https://docs.rs/s2n-quic/latest/s2n_quic/connection/struct.Connection.html) which can do more interesting stuff.

Having separate  pre-connect/post-connect struct/state is important to prevent someone from trying to `poll_until(state)` before establishing a connection with the peer.

I have added some documentation to this and renamed it to `Builder`.

File: netbench-orchestrator/README.md [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+tasks such as installing/updating dependencies and uploading netbench data to s3. SSM is
+also used to start the Russula worker process on each host. Russula is primarily used for
+timing sensitive tasks such netbench, which require starting and stopping multiple servers
+and clients processes across multiple hosts. Specifically, the 'Worker' component of a
+workflow is executed on the remote host, while the 'Coordinator' component is run locally
+as part of the Orchestrator.
+
+### Debugging
+As discussed in the above overview, there are processes that run locally and those that run
+remotely. This sections describes how to go about debugging each component.
```
Comment (WesleyRosenblum) [inline netbench-orchestrator/README.md:47]:
```suggestion
remotely. This section describes how to go about debugging each component.
```

**Review by WesleyRosenblum — APPROVED** [review]
